src/context.js

```diff
@@ -0,0 +1,54 @@
+// @flow
+import * as React from 'react';
+import {createPlugin} from 'fusion-core';
+import type FusionApp, {FusionPlugin, Middleware, Token} from 'fusion-core';
+import type {Element} from 'react';
+
+type ExtractReturnType = <V>(() => V) => V;
+
+type ServiceContextType<T> = {
+  (): $Call<ExtractReturnType, T>,
+};
+
+export const ServiceContext = React.createContext<ServiceContextType<any>>(
+  () => {}
+);
+
+export function useService<T: Token<any>>(token: T): ServiceContextType<T> {
```

Comment:
Are we able to avoid the usage of `any` here?  With @ganemone's suggestion above to avoid explicitly typing the generic, I think we can improve this.

Maybe something like this:

```
export function useService<TService>(
  token: Token<TService>
): ServiceContextType<TService> {
  const getService: (
    Token<TService>
  ) => ServiceContextType<TService> = React.useContext(ServiceContext);
  const provides = getService(token);
  return provides;
}
```

We could remove the usage for `ExtractReturnType` entirely then, and `ServiceContextType` would just return `T` directly:

```
type ServiceContextType<TService> = {
   (): TService,
 };
```

Reply:
Yeah that makes sense. Now that it's been cleared up that you don't need to pass the type into `useService` this will be easier. With this change, is it necessary to have the `ServiceContextType` at all? It just looks like an identity type now

Reply:
Yea that looks good assuming it works like we think.

Reply:
Hm, we may even be able to do away with `ServiceContextType` here as `provides` should resolve to the actual service itself (as opposed to an invocable object that resolves to the service).